progress test fail

I tested already the progress api and it works as expected

I tried to checkout the repo, but from some reason this test fails https://github.com/dsherret/dax/blob/main/mod.test.ts#L881

```
progress => ./mod.test.ts:881:6
error: AssertionError: Values are not equal:


    [Diff] Actual / Expected


    [
+     "Downloading Test",
    ]

  throw new AssertionError(message);
        ^
    at assertEquals (https://deno.land/std@0.167.0/testing/asserts.ts:190:9)
    at file:///home/mrcool/dev/deno/others/dax/mod.test.ts:886:3
```

I followed with a debugger and it seems for some reason the instantiation seems to fail here https://github.com/dsherret/dax/blob/main/src/console/utils.ts#L166 , so its weird I get the assertion error then, how did it reach there , maybe assertEquals has some internal try catch

I made sure I regenerated the wasm file with deno task wasmbuild, but same test error

--- a/README.md
+++ b/README.md
@@ -396,16 +396,16 @@ await pb.with(async () => {
 
 #### Synchronous work
 
-The progress bars are updated on an interval (via `setInterval`). If you are doing a lot of synchronous work, it will start updating the progress bar on the current execution context, but this only occurs in some cases. Due to this, it's probably best to force a render where you think it would be appropriate by using the `.forceRender()` method:
+The progress bars are updated on an interval (via `setInterval`). If you are doing a lot of synchronous work the progress bars won't update. Due to this, you can force a render where you think it would be appropriate by using the `.forceRender()` method:
 
 ```ts
 const pb = $.progress("Processing Items");
 
-pb.with(() => {
+await pb.with(async () => {
   for (const item of items) {
-    doWork(item); // note this does not use `await` so we force rendering below
+    doWork(item);
     pb.increment();
-    pb.forceRender();
+    await pb.forceRender();
   }
 });
 ```

--- a/mod.test.ts
+++ b/mod.test.ts
@@ -878,11 +878,11 @@ Deno.test("pwd: pwd", async () => {
   assertEquals(await $`pwd`.text(), Deno.cwd());
 });
 
-Deno.test("progress", () => {
+Deno.test("progress", async () => {
   const logs: string[] = [];
   $.setInfoLogger((...data) => logs.push(data.join(" ")));
   const pb = $.progress("Downloading Test");
-  pb.forceRender(); // should not throw;
+  await pb.forceRender(); // should not throw;
   assertEquals(logs, [
     "Downloading Test",
   ]);

--- a/src/console/logger.ts
+++ b/src/console/logger.ts
@@ -12,12 +12,14 @@ const refreshItems: Record<LoggerRefreshItemKind, TextItem[] | undefined> = {
 
 function setItems(kind: LoggerRefreshItemKind, items: TextItem[] | undefined, size?: ConsoleSize) {
   refreshItems[kind] = items;
-  return refresh(size);
+  refresh(size);
 }
 
-async function refresh(size?: ConsoleSize) {
-  const staticText = await getStaticText();
-  refreshWithStaticText(staticText, size);
+function refresh(size?: ConsoleSize) {
+  const staticText = getStaticTextIfCreated();
+  if (staticText) {
+    refreshWithStaticText(staticText, size);
+  }
 }
 
 function refreshWithStaticText(staticText: Awaited<ReturnType<typeof getStaticText>>, size?: ConsoleSize) {
@@ -44,7 +46,17 @@ async function logOnce(items: TextItem[], size?: ConsoleSize) {
   }, size);
 }
 
+async function ensureInitialized() {
+  await getStaticText();
+}
+
+function isInitilaized() {
+  return getStaticTextIfCreated() != null;
+}
+
 const logger = {
+  ensureInitialized,
+  isInitilaized,
   setItems,
   logOnce,
   logAboveStaticText,

--- a/src/console/progress/interval.ts
+++ b/src/console/progress/interval.ts
@@ -1,5 +1,5 @@
 import { logger, LoggerRefreshItemKind } from "../logger.ts";
-import { ConsoleSize, isInteractiveConsole, TextItem } from "../utils.ts";
+import { ConsoleSize, isOutputTty, TextItem } from "../utils.ts";
 
 export interface RenderIntervalProgressBar {
   render(size: ConsoleSize): TextItem[];
@@ -8,15 +8,16 @@ export interface RenderIntervalProgressBar {
 const intervalMs = 60;
 const progressBars: RenderIntervalProgressBar[] = [];
 let renderIntervalId: number | undefined;
-let lastRenderTime = Date.now();
 
 export function addProgressBar(render: (size: ConsoleSize) => TextItem[]): RenderIntervalProgressBar {
   const pb = {
     render,
   };
   progressBars.push(pb);
-  if (renderIntervalId == null && isInteractiveConsole) {
-    renderIntervalId = setInterval(forceRender, intervalMs);
+  if (renderIntervalId == null && isOutputTty) {
+    const _ignore = logger.ensureInitialized()
+      .catch(() => {/* ignore */});
+    renderIntervalId = setInterval(forceRenderSync, intervalMs);
   }
   return pb;
 }
@@ -35,24 +36,31 @@ export function removeProgressBar(pb: RenderIntervalProgressBar) {
   return true;
 }
 
-export function forceRenderIfHasNotInWhile() {
-  const duration = Date.now() - lastRenderTime;
-  if (duration > intervalMs * 2) {
-    forceRender();
+export function forceRender() {
+  if (!isShowingProgressBars()) {
+    return;
+  }
+
+  if (logger.isInitilaized()) {
+    forceRenderSync();
+    return Promise.resolve();
+  } else {
+    return logger.ensureInitialized().then(() => {
+      forceRenderSync();
+    });
   }
 }
 
-export function forceRender() {
-  if (!isInteractiveConsole || progressBars.length === 0) {
+function forceRenderSync() {
+  if (!isShowingProgressBars()) {
     return;
   }
 
   const size = Deno.consoleSize();
   const items = progressBars.map((p) => p.render(size)).flat();
   logger.setItems(LoggerRefreshItemKind.ProgressBars, items, size);
-  lastRenderTime = Date.now();
 }
 
 export function isShowingProgressBars() {
-  return isInteractiveConsole && progressBars.length > 0;
+  return isOutputTty && progressBars.length > 0;
 }

--- a/src/console/progress/mod.ts
+++ b/src/console/progress/mod.ts
@@ -1,13 +1,7 @@
 import { colors } from "../../deps.ts";
-import { ConsoleSize, isInteractiveConsole, safeConsoleSize, TextItem } from "../utils.ts";
+import { ConsoleSize, isOutputTty, safeConsoleSize, TextItem } from "../utils.ts";
 import { humanDownloadSize } from "./format.ts";
-import {
-  addProgressBar,
-  forceRender,
-  forceRenderIfHasNotInWhile,
-  removeProgressBar,
-  RenderIntervalProgressBar,
-} from "./interval.ts";
+import { addProgressBar, forceRender, removeProgressBar, RenderIntervalProgressBar } from "./interval.ts";
 
 export { isShowingProgressBars } from "./interval.ts";
 
@@ -62,7 +56,6 @@ export class ProgressBar {
   /** Sets the prefix message/word, which will be displayed in green. */
   prefix(prefix: string | undefined) {
     this.#state.prefix = prefix;
-    forceRenderIfHasNotInWhile();
     if (prefix != null && prefix.length > 0) {
       this.#logIfNonInteractive();
     }
@@ -72,7 +65,6 @@ export class ProgressBar {
   /** Sets the message the progress bar will display after the prefix in white. */
   message(message: string | undefined) {
     this.#state.message = message;
-    forceRenderIfHasNotInWhile();
 
     if (message != null && message.length > 0) {
       this.#logIfNonInteractive();
@@ -87,7 +79,7 @@ export class ProgressBar {
   }
 
   #logIfNonInteractive() {
-    if (isInteractiveConsole) {
+    if (isOutputTty) {
       return;
     }
     let text = this.#state.prefix ?? "";
@@ -103,21 +95,18 @@ export class ProgressBar {
   /** Sets the current position of the progress bar. */
   position(position: number) {
     this.#state.currentPos = position;
-    forceRenderIfHasNotInWhile();
     return this;
   }
 
   /** Increments the position of the progress bar. */
   increment(inc = 1) {
     this.#state.currentPos += inc;
-    forceRenderIfHasNotInWhile();
     return this;
   }
 
   /** Sets the total length of the progress bar. */
   length(size: number | undefined) {
     this.#state.length = size;
-    forceRenderIfHasNotInWhile();
     return this;
   }
 
@@ -129,7 +118,7 @@ export class ProgressBar {
 
   /** Forces a render to the console. */
   forceRender() {
-    forceRender();
+    return forceRender();
   }
 
   /** Finish showing the progress bar. */
@@ -145,39 +134,16 @@ export class ProgressBar {
     }
   }
 
-  /**
-   * Does the provided action and will call `.finish()` when this is the last `.with(...)` action that runs.
-   *
-   * Since this is the synchronous overload, you should probably call `.forceRender()` to force rendering
-   * every now and again.
-   */
-  with<TResult>(action: () => TResult): TResult;
   /** Does the provided action and will call `.finish()` when this is the last `.with(...)` action that runs. */
-  with<TResult>(action: () => Promise<TResult>): Promise<TResult>;
-  with<TResult>(action: () => Promise<TResult> | TResult) {
+  async with<TResult>(action: () => Promise<TResult>): Promise<TResult> {
     this.#withCount++;
-    let wasAsync = false;
     try {
-      const result = action();
-      if (result instanceof Promise) {
-        wasAsync = true;
-        return result.finally(() => {
-          this.#decrementWithCount();
-        });
-      } else {
-        return result;
-      }
+      return await action();
     } finally {
-      if (!wasAsync) {
-        this.#decrementWithCount();
+      this.#withCount--;
+      if (this.#withCount === 0) {
+        this.finish();
       }
-    }
-  }
-
-  #decrementWithCount() {
-    this.#withCount--;
-    if (this.#withCount === 0) {
-      this.finish();
     }
   }
 }

--- a/src/console/utils.ts
+++ b/src/console/utils.ts
@@ -68,13 +68,8 @@ export function showCursor() {
   Deno.stderr.writeSync(encoder.encode("\x1B[?25h"));
 }
 
-export function ensureTty(title: string) {
-  if (!Deno.isatty(Deno.stdin.rid)) {
-    throw new Error(`Cannot prompt when not a tty. (Prompt: '${title}')`);
-  }
-}
-
-export const isInteractiveConsole = Deno.isatty(Deno.stdin.rid) && safeConsoleSize() != null;
+const canGetConsoleSize = safeConsoleSize() != null;
+export const isOutputTty = canGetConsoleSize && Deno.isatty(Deno.stderr.rid);
 
 export function resultOrExit<T>(result: T | undefined): T {
   if (result == null) {
@@ -92,27 +87,30 @@ export interface SelectionOptions<TReturn> {
 }
 
 export function createSelection<TReturn>(options: SelectionOptions<TReturn>): Promise<TReturn | undefined> {
-  ensureTty(options.message);
+  if (!isOutputTty || !Deno.isatty(Deno.stdin.rid)) {
+    throw new Error(`Cannot prompt when not a tty. (Prompt: '${options.message}')`);
+  }
   if (safeConsoleSize() == null) {
     throw new Error(`Cannot prompt when can't get console size. (Prompt: '${options.message}')`);
   }
   return ensureSingleSelection(async () => {
-    await logger.setItems(LoggerRefreshItemKind.Selection, options.render());
+    await logger.ensureInitialized();
+    logger.setItems(LoggerRefreshItemKind.Selection, options.render());
 
     for await (const key of readKeys()) {
       const keyResult = options.onKey(key);
       if (keyResult != null) {
         const size = Deno.consoleSize();
-        await logger.setItems(LoggerRefreshItemKind.Selection, [], size);
+        logger.setItems(LoggerRefreshItemKind.Selection, [], size);
         if (options.noClear) {
           await logger.logOnce(options.render(), size);
         }
         return keyResult;
       }
-      await logger.setItems(LoggerRefreshItemKind.Selection, options.render());
+      logger.setItems(LoggerRefreshItemKind.Selection, options.render());
     }
 
-    await logger.setItems(LoggerRefreshItemKind.Selection, []); // clear
+    logger.setItems(LoggerRefreshItemKind.Selection, []); // clear
     return undefined;
   });
 }

--- a/src/console/utils.ts
+++ b/src/console/utils.ts
@@ -68,8 +68,7 @@ export function showCursor() {
   Deno.stderr.writeSync(encoder.encode("\x1B[?25h"));
 }
 
-const canGetConsoleSize = safeConsoleSize() != null;
-export const isOutputTty = canGetConsoleSize && Deno.isatty(Deno.stderr.rid);
+export const isOutputTty = safeConsoleSize() != null && Deno.isatty(Deno.stderr.rid);
 
 export function resultOrExit<T>(result: T | undefined): T {
   if (result == null) {

